### ISSUE TITLE
Can't use css variables in $success and other state sass variables.

### ISSUE THREAD

**Issue by ethib137** [liferay/clay#4306]
I want to be able to use style books to set success, info, warning, and error colors but the theme will not compile. 

`$success: var(--color-state-success);`

**Comment by pat270** [comment]
@ethib137 send a pr for this, but you need to have a fallback value for your CSS variable (e.g., `$success: var(--color-state-success, #4aab3b);`). See https://github.com/liferay/clay/pull/4309#issue-1008719849

**Comment by pat270** [comment]
Copy pasting comment from https://github.com/liferay/clay/pull/4309#issuecomment-928909687 to keep discussion in one place.

> @pat270 it looks like any time one of the sass functions to modify the color is used, the css custom property will be replaced by it's fallback:
```
@function clay-darken($color, $amount) {
	$color: clay-get-fallback($color);
	$amount: clay-get-fallback($amount);

	@return darken($color, $amount);
}
```
> That would mean that if I change the value of the custom property it won't update in places that use the color sass functions because the fallback value hasn't changed. Is that right?

> If so, it makes using a css custom property ineffective, since it's dynamic changes will not always be applied. :(

> So, what are our other options?

> My preference would be that we move away from using sass functions to modify colors. Whenever we do this it limits how much we can customize the look and feel for the end user. I realize this is likely impractical to do in the short term, so I would propose an alternative:

> Whenever we use a sass function to modify color the result should be set to a sass variable before being applied to style a selector, so that it's easy to override the use of the function. As an example...

> Do this:
```
$danger-d1: darken($danger, 5) !default;
```
> Not this:
```
hr {
	border-top-color: darken(map-get($cadmin-value, border-color), 5%);
}
```
> I realize this wouldn't actually solve the original problem I brought up, but it would make sure that a work around exists.

> What do you think?


@ethib137 Yes colors will be static. If you change `--color-state-success: red;`, everything using `--color-state-success` will be red, but the success component hover and focus states (for example) will be green.

A majority of the Sass color function stuff is from Bootstrap. It doesn't work well unless you pick similar shades (don't know the correct term) as Bootstrap's defaults for primary, secondary, success, etc. The place where Clay intentionally uses Sass color functions are in the theme color variables:

```scss
$primary: #0b5fff !default;
$primary-d1: clay-darken($primary, 5.1) !default;
$primary-d2: clay-darken($primary, 10) !default;
$primary-l1: clay-lighten($primary, 22.94) !default;
$primary-l2: clay-lighten($primary, 32.94) !default;
$primary-l3: clay-lighten($primary, 44.9) !default;
```

It was requested by Lexicon. I don't even like using them there, we should plug in the hex values directly.

For dynamic colors via CSS Custom Properties, plug in custom properties / values you need for each color variant? You will need to get creative with it and find a CSS only solution that fits your needs. Material Design uses pseudo elements as an overlay (white or black) with opacity to create darker and lighter shades for hover, focus, active states. Maybe CSS filter, rgba, hsl?

**Comment by matuzalemsteles** [comment]
Hey @pat270 I think I understand the @ethib137 question here, I think the problem he said is more related, the sass variables that are using `clay-lighten` or `clay-darken`... despite having a fallback which will "support" CSS variables will not be used, because they are used at build time, so passing a fallback just makes things compile but don't replace it with a CSS variable but keep keeping the fallback.

I think this makes sense, considering that you want a sass variable to use a CSS variable instead of having a fallback, it just masks the compilation error by the native properties of sass not accepting CSS variable. In these cases, I think the fallback instead of always returning the result of the `lighten` call if it is a CSS variable returns it.

**Comment by pat270** [comment]
For the `hr` thing, I just sent a pr fixing that at https://github.com/liferay/clay/pull/4308. You can do:

```scss
$alert-success: (
  hr: (
    border-top-color: var(--color-state-success),
    opacity: 0.2,
  ),
);
```
or

```scss
$alert-success: (
  hr: (
    border-top-color: var(--color-state-success-hr-border-top-color),
  ),
);
```

What was there previously is Bootstrap's [alert-variant](https://github.com/liferay/clay/blob/27032c4b83b59318ccdacaaf7627e80de277726a/packages/clay-css/src/scss/mixins/_alerts.scss#L1) mixin modified. I modified the minimum until we had the infrastructure in place to do what we want.

Keep keeping the fallback is the only way we can accommodate CSS variables with Sass color functions. They are inherited from Bootstrap and I don't want to break it until 4.x. The solution is to keep converting components to that pattern. It allows us to satisfy everything.

**Comment by ethib137** [comment]
Yeah I agree with @matuzalemsteles... I think a better fallback for the darken and lighten functions would be to just return the css custom property, not the fallback.

The fix for alert is good, but there are many other instances that also need to be changed:

- liferay/clay/packages/clay-css/src/scss/bootstrap/mixins/_background-variant.scss 12
- liferay/clay/packages/clay-css/src/scss/bootstrap/mixins/_badge.scss
- liferay/clay/packages/clay-css/src/scss/bootstrap/mixins/_list-group.scss
- etc... Anywhere else a sass function is applied directly to a style without the ability to override it.

**Comment by pat270** [comment]
I can meet you guys half way. I'll add a feature flag like `$enable-return-css-variable-from-color-functions` (need a better name) where it will return the custom property as is. We can enable it by default and give a way out if someone doesn't like it.

**Comment by matuzalemsteles** [comment]
@pat270 makes sense to me, I think it might be a good fit.

**Comment by ethib137** [comment]
@pat270 yeah I like this option too!
